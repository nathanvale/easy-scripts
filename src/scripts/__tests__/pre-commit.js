/* eslint-disable jest/no-standalone-expect */
import cases from "jest-in-case";
import {
  unquoteSerializer,
  winPathSerializer,
} from "../../helpers/serializers";

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);

describe("scripts", () => {
  cases(
    "pre-commit",
    ({
      args = [],
      utils = require("../../utils"),
      hasPkgProp = () => false,
      hasFile = () => false,
    }) => {
      // beforeEach
      const { sync: crossSpawnSyncMock } = require("cross-spawn");
      const originalArgv = process.argv;
      const originalExit = process.exit;
      Object.assign(utils, {
        hasPkgProp,
        hasFile,
        resolveBin: (modName, { executable = modName } = {}) => executable,
      });
      jest.spyOn(process, "exit").mockImplementation();

      process.argv = ["node", "../pre-commit", ...args];
      utils.isOptedIn = (optIn) => optIn === "pre-commit";

      try {
        // tests
        require("../pre-commit");
        expect(crossSpawnSyncMock).toHaveBeenCalledTimes(2);
        const [firstCall, secondCall] = crossSpawnSyncMock.mock.calls;
        const [scriptOne, calledArgsOne] = firstCall;
        expect([scriptOne, ...calledArgsOne].join(" ")).toMatchSnapshot();
        const [scriptTwo, calledArgsTwo] = secondCall;
        expect([scriptTwo, ...calledArgsTwo].join(" ")).toMatchSnapshot();
        // eslint-disable-next-line no-useless-catch
      } catch (error) {
        throw error;
      } finally {
        // afterEach
        process.exit = originalExit;
        process.argv = originalArgv;
        jest.resetModules();
      }
    },
    {
      "calls lint-staged CLI with default args": {},
      "does not use built-in config with --config": {
        args: ["--config", "./custom-config.js"],
      },
      "does not use built-in config with .lintstagedrc file": {
        hasFile: (filename) => filename === ".lintstagedrc",
      },
      "does not use built-in config with lint-staged.config.js file": {
        hasFile: (filename) => filename === "lint-staged.config.js",
      },
      "does not use built-in config with lint-staged pkg prop": {
        hasPkgProp: (prop) => prop === "lint-staged",
      },
      "forwards args": {
        args: ["--verbose"],
      },
    }
  );

  test("should exit if lintstaged fails ", () => {
    const { sync: crossSpawnSyncMock } = require("cross-spawn");
    const originalExit = process.exit;
    const status = { status: 1 };
    crossSpawnSyncMock.mockReturnValueOnce(status);
    jest.spyOn(process, "exit").mockImplementation();
    require("../pre-commit");
    expect(process.exit).toBeCalledWith(1);
    process.exit = originalExit;
  });
});
