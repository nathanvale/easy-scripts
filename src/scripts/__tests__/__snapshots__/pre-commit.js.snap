// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scripts pre-commit calls lint-staged CLI with default args 1`] = `lint-staged --config ./src/config/lintstagedrc.js`;

exports[`scripts pre-commit calls lint-staged CLI with default args 2`] = `npm run validate`;

exports[`scripts pre-commit does not use built-in config with .lintstagedrc file 1`] = `lint-staged`;

exports[`scripts pre-commit does not use built-in config with .lintstagedrc file 2`] = `npm run validate`;

exports[`scripts pre-commit does not use built-in config with --config 1`] = `lint-staged --config ./custom-config.js`;

exports[`scripts pre-commit does not use built-in config with --config 2`] = `npm run validate`;

exports[`scripts pre-commit does not use built-in config with lint-staged pkg prop 1`] = `lint-staged --config ./src/config/lintstagedrc.js`;

exports[`scripts pre-commit does not use built-in config with lint-staged pkg prop 2`] = `npm run validate`;

exports[`scripts pre-commit does not use built-in config with lint-staged.config.js file 1`] = `lint-staged`;

exports[`scripts pre-commit does not use built-in config with lint-staged.config.js file 2`] = `npm run validate`;

exports[`scripts pre-commit forwards args 1`] = `lint-staged --config ./src/config/lintstagedrc.js --verbose`;

exports[`scripts pre-commit forwards args 2`] = `npm run validate`;
